File: lua/xbase/broadcast.lua
local logger = require "xbase.logger"
local notify = require "xbase.notify"
local socket = require "xbase.socket"
local types = require "xbase.types"
local msg, tkind, tstatus = types.Message, types.TaskKind, types.TaskStatus
local config = require("xbase.config").values

local M = {}

M.expect_second_run = false

local function task_set(args)
  M.has_task = true
  local running, _ = tkind:prefix(args.kind)
  args.prefix = running
  vim.g.xbase_ctask = args
  local line = string.format("[%s] %s", args.target, args.prefix)
  if tkind.is_run(args.kind) then
    vim.g.xbase_ctask_line = string.format("%s %s", config.statusline.device_running.icon, line)
  else
    vim.g.xbase_ctask_line = line
  end
  vim.schedule(require("xbase.statusline").update_spinner)
end

local function task_update(args)
  local content, level = args.content, args.level
  if #content == 0 then
    return
  end
  local ctask = vim.g.xbase_ctask
  local target, prefix, kind = ctask.target, ctask.prefix, ctask.kind

  logger.log(content, level)

  if level ~= "Debug" and level ~= "Trace" then
    content = string.gsub(content, "%[" .. target .. "%]%s", "")
    local line = string.format("[%s] %s: %s", target, prefix, content)

    if tkind.is_run(kind) then
      local icon = config.statusline.device_running.icon
      line = ("%s %s"):format(icon, line)
      vim.g.xbase_ctask_display = line
    end

    vim.g.xbase_ctask_line = line
  end
end

local function task_finish(args)
  M.has_task = false
  vim.g.xbase_ctask = vim.tbl_extend("force", vim.g.xbase_ctask, args)
  local line, level, icon
  local ctask = vim.g.xbase_ctask
  local target, prefix = ctask.target, ctask.prefix
  local is_failed = tstatus.is_failed(args.status)

  if is_failed then
    icon = config.statusline.failure.icon
    level = "Error"
  else
    icon = config.statusline.success.icon
    level = "Info"
  end

  if tkind.is_run(vim.g.xbase_ctask.kind) then
    vim.g.xbase_ctask_display = ("%s [%s] Device Disconnected"):format(icon, target)
  else
    if is_failed then
      line = ("[%s] %s Failed"):format(target, prefix)
    else
      local _, done = tkind:prefix(ctask.kind)
      line = ("[%s] %s"):format(target, done)
    end
    vim.g.xbase_ctask_line = ("%s %s"):format(icon, line)
  end

  vim.schedule(function()
    if line then
      logger.log(line, level)
    end
    vim.defer_fn(function()
      if M.has_task == false then
        vim.g.xbase_ctask_display = nil
        vim.g.xbase_ctask_line = nil
      end
    end, 5000)
  end)
end

function M.start(root, address)
  local socket = socket:connect(address)

  socket._socket:write(string.format("%s\n", vim.loop.os_getpid()), function(err)
    if err then
      print(socket._stream_error or err)
    end
  end)

  socket:read_start(function(chunk)
    local chunk = vim.trim(chunk)
    for _, chunk in ipairs(vim.split(chunk, "\n")) do
      local item = vim.json.decode(chunk)
      local type, args = item.type, item.args

      vim.schedule(function()
        if msg.task_is_update_current(type) then
          return task_update(args)
        end

        if msg.task_is_set_current(type) then
          return task_set(args)
        end

        if msg.task_is_finish_current(type) then
          return task_finish(args)
        end

        if msg.is_notify(type) then
          notify(args.content, args.level)
          if string.find(args.content, "Registered") ~= nil then
            vim.schedule(function()
              vim.defer_fn(function()
                print "  "
              end, 2000)
            end)
          end
          return
        end

        if msg.is_reload_lsp_server(type) then
          return vim.cmd "LspRestart"
        end

        if msg.is_open_logger(type) then
          return logger.toggle(nil, false)
        end

        if msg.is_log(type) then
          return logger.log(args.content, args.level)
        end

        if msg.is_set_state(type) then
          local key, value = args.key, args.value
          if key == "runners" then
            require("xbase.state").runners = value
          elseif key == "projectInfo" then
            require("xbase.state").project_info[root] = value
          end
          return
        end

        ---@diagnostic disable-next-line: empty-block
        if msg.is_set_watching(type) then
          -- ()
        end
      end)
    end
  end)

  return socket
end

return M

File: vscode/xbase/broadcast.ts
import { filter, map, pipe, split, toAsync } from "iter-ops";
import net from "net";
import { Disposable, window, commands } from "vscode";
import { Message, ContentLevel, TaskKind, TaskStatus } from "./types";
import Logger from "./ui/logger";
import Statusline from "./ui/statusline";
import configuration from "./config";
import { WorkspaceContext } from "./workspaceContext";
import SourcekitLsp from "./sourcekit-lsp";
import FolderContext from "./folderContext";

interface CurrentTask {
  prefix: { processing: string, done: string },
  target: string,
  kind: TaskKind,
  status: TaskStatus,
}

export default class Broadcast implements Disposable {
  public name: string;
  public folderCtx: FolderContext;
  public socket: net.Socket;
  private logger: Logger;
  private statusline: Statusline;
  private currentTask?: CurrentTask;
  private sourcekit: SourcekitLsp;

  private constructor(
    folder: FolderContext, socket: net.Socket, ctx: WorkspaceContext
  ) {
    this.folderCtx = folder;
    this.name = folder.name.charAt(0).toUpperCase() + folder.name.slice(1);
    this.socket = socket;
    this.logger = ctx.logger;
    this.statusline = ctx.statusline;
    this.sourcekit = ctx.sourcekit;
  }

  public static async connect(
    folder: FolderContext,
    address: string,
    ctx: WorkspaceContext
  ): Promise<Broadcast> {
    return new Promise((resolve, reject) => {
      const socket = net.createConnection(address, () => {
        const broadcast = new Broadcast(folder, socket, ctx);
        socket.on("data", async buffer => {
          for await (const message of Broadcast.get_messages(buffer))
            await broadcast.handleMessage(message);
        });
        socket.write(`${process.pid}\n`);
        resolve(broadcast);
      });
      socket.on("error", err => {
        reject(Error(`Failed to connect to XBase Broadcast: ${err}`));
      });
    });
  }

  private static get_messages(buffer: Buffer): AsyncIterable<Message> {
    return pipe(
      toAsync(buffer),
      split(a => a === 10),
      map(m => Buffer.from(m)),
      filter(m => m.length > 1),
      map(m => JSON.parse(m.toString()) as Message)
    );
  }

  private async handleMessage(message: Message) {
    switch (message.type) {
      case "Notify": {
        const { content, level } = message.args;
        await this.notify(content, level);
        break;
      }
      case "Log": {
        const { content, level } = message.args;
        // TODO: check for current log level
        if (!levelShouldIgnore(level))
          this.logger.append(content, level);
        break;
      }
      case "OpenLogger":
        if (configuration.ui.openLoggerOnError)
          this.logger.toggle();
        break;
      case "SetCurrentTask":
        this.setTask(message.args.kind, message.args.target, message.args.status);
        break;
      case "UpdateCurrentTask":
        this.updateCurrentTask(message.args.content, message.args.level);
        break;
      case "FinishCurrentTask":
        await this.finishTask(message.args.status);
        break;
      case "ReloadLspServer":
        await this.sourcekit.restartClient(this.folderCtx.uri);
        break;
      case "SetState":
        switch (message.args.key) {
          case "runners":
            console.log("Runners are set");
            this.folderCtx.ctx.runners = message.args.value;
            break;
          case "projectInfo":
            console.log("projectInfo is set");
            this.folderCtx.projectInfo = message.args.value;
            break;
        }
        break;
    }
  }

  private setTask(kind: TaskKind, target: string, status: TaskStatus) {
    const prefix = TaskKind.prefix(kind)!;
    this.currentTask = { target, kind, status, prefix };
    this.statusline.update({
      content: `[${target}] ${prefix.processing}`,
      icon: TaskKind.isRun(kind) ? "$(code)" : undefined
    });
  }

  private updateCurrentTask(content: string, level: ContentLevel) {
    if (this.currentTask === undefined) {
      console.warn("trying to update task that no longer exists!");
      return;
    };

    if (levelShouldIgnore(level))
      return;

    const { target, prefix, kind } = this.currentTask;

    this.logger.append(content, level);

    content = content.replace(`[${this.currentTask.target}]`, "");

    this.statusline.update({
      content: `[${target}] ${prefix.processing}: ${content}`,
      icon: TaskKind.isRun(kind) ? "$(code)" : undefined,
      level
    });
  }

  private async finishTask(status: TaskStatus) {
    if (this.currentTask === undefined) {
      console.warn("trying to finish task that no longer exists!");
      return;
    }

    const { target, prefix, kind } = { ...this.currentTask };
    const taskFailed = (status === "Failed");
    this.currentTask = undefined;

    const level = taskFailed ? "Error" : "Info";
    const content = TaskKind.isRun(kind)
      ? `[${target}] Device disconnected`
      : (taskFailed
        ? `[${target}] ${prefix.processing} Failed`
        : `[${target}] ${prefix.done}`);

    this.logger.append(content, level);

    this.statusline.set({
      icon: taskFailed ? "$(error)" : "$(pass)",
      content,
      level,
      is_success: taskFailed ? false : true
    });


    if (!taskFailed) {
      await (new Promise(resolve => setTimeout(resolve, 3000)));
      this.statusline.setDefault();
    }
  }

  private async notify(msg: string, level: ContentLevel) {
    switch (level) {
      case "Info":
        window.showInformationMessage(msg);
        if (msg.trim().endsWith("Registered")) {
          setTimeout(async () => {
            console.log("Clearing ...");
            await commands.executeCommand("notifications.clearAll");
          }, 3500);
        }
        break;
      case "Warn":
        window.showWarningMessage(msg);
        break;
      case "Error":
        window.showErrorMessage(msg);
        break;
    }
  }

  dispose() {
    this.socket.pause();
    this.socket.end();
    this.socket.destroy();
  }
}

const levelShouldIgnore = (level: ContentLevel) =>
  (level === "Debug" || level === "Trace");
